Securing a Graphql API (#280)
neo4j/docs-graphql: modules/ROOT/pages/security/securing-a-graphql-api.adoc

[[securing-an-api]]
:description: This page is a tutorial on how to secure your API created with the Neo4j GraphQL Library.
= Securing your GraphQL API

This page is a tutorial on how to secure your API created with the Neo4j GraphQL Library.

== Prerequisites

. Set up a new AuraDB instance.
Refer to link:https://neo4j.com/docs/aura/getting-started/create-instance/[Creating a Neo4j Aura instance].
. Populate the instance with the Northwind data set.

[NOTE]
====
If you have completed the GraphQL and Aura Console getting started guide and would like to get rid of the example nodes you have created there, run the following in **Query** before populating your data base with the Northwind set:
[source,cypher]
----
MATCH (n) DETACH DELETE n;
----
====

This tutorial builds on top of the xref:graphql-modeling.adoc[GraphQL modeling tutorial].
Specifically, it extends the following type definitions:

[source, graphql, indent=0]
----
type Customer @node {
    contactName: String!
    customerID: ID! @id
    orders: [Order!]! @relationship(type: "PURCHASED", direction: OUT)
}

> type Order @limit(max: 20, default: 10) @node {

Author: ```suggestion
type Order @limit(max: 10, default: 10) @node {
```

}
----

The `sensitiveData` field is neither available for queries nor for subscriptions nor for aggregations.
The `createdAt` field can be set when a new customer is created, but it cannot be updated.

=== Full example

Here is the full set of type definitions extended with security-related directives:

[source, graphql, indent=0]
----
type JWT @jwt {
    roles: [String!]!
    customerID: String! @jwtClaim(path: "sub")
}

type Customer
    @node
    @authentication(operations: [DELETE], jwt: { roles: { includes: "admin" } })
    @authorization(
        filter: [
            { operations: [READ], where: { node: { customerID: { eq: "$jwt.customerID" } } } }
            { where: { jwt: { roles: { includes: "admin" } } } }
        ]
    ) {
    contactName: String!
    sensitiveData: String! @selectable(onRead: false, onAggregate: false)
    createdAt: DateTime! @settable(onCreate: true, onUpdate: false)
    adminNotes: [String!]! @authorization(validate: [{ where: { jwt: { roles: { includes: "admin" } } } }])
    customerID: ID! @id
    orders: [Order!]! @relationship(type: "PURCHASED", direction: OUT)
}

type Order

>     @limit(max: 20, default: 10) 

Author: ```suggestion
    @limit(max: 10, default: 10) 
```

[[securing-an-api]]
:description: This page is a tutorial on how to secure your API created with the Neo4j GraphQL Library.
= Securing your GraphQL API

This page is a tutorial on how to secure your API created with the Neo4j GraphQL Library.

== Prerequisites

. Set up a new AuraDB instance.
Refer to link:https://neo4j.com/docs/aura/getting-started/create-instance/[Creating a Neo4j Aura instance].

> . Populate the instance with the Northwind data set.

Reviewer: Are we sure we want to make this a pre-requisite? This can also be done with an on-prem db

Author: it's the same in graphql-modeling.adoc, following the "aura first" approach (making aura the default UX journey)

    orders: [Product!]! @relationship(type: "ORDERS", direction: IN, properties: "ordersProperties")
    supplier: [Supplier!]! @relationship(type: "SUPPLIES", direction: IN)
}
type Category @node {
    categoryName: String!
    products: [Product!]! @relationship(type: "PART_OF", direction: IN)
}
type Supplier @node {
    supplierID: ID! @id
    companyName: String!
    products: [Product!]! @relationship(type: "SUPPLIES", direction: OUT)
}
type ordersProperties @relationshipProperties {
    unitPrice: Float!
    quantity: Int!
}
----

== Security-related directives

The GraphQL Library has several directives dedicated to security: xref:security/authentication.adoc[`@authentication`] and xref:security/authorization.adoc[`@authorization`], as well as xref:security/configuration.adoc#_jwt[`@jwt`] and xref:security/configuration.adoc#_jwtclaim[`@jwtClaim`].
The xref:directives/schema-configuration/field-configuration.adoc#_selectable[`@selectable`] and xref:directives/schema-configuration/field-configuration.adoc#_settable[`@settable`] directives can be used to control accessibility of data fields through certain operations.

=== Authentication

> The xref:security/authentication.adoc[`@authentication` directive] can be applied globally, only to certain fields or only to certain types, and only for certain operations.

Reviewer: can be applied globally but only to certain elements?

Is there a typo? I'm not sure I understand this

Author: rephrased
it's either ... or ... or (added "either" and made it active voice)

> Add admin authorization for operations on customers, orders, products, categories and suppliers:

Reviewer: We use different directives to differentiate between authentication (we can identify who you are) and authorization (you have the correct rights to access the data)

While this particular example is a bit on the edge, I would avoid using the word "authorization" when trying to add examples of authentication to avoid confusion

In this case, we are validating your tokens are valid, including the role

Author: rephrased a bit

* `DELETE` for customers,
* `UPDATE` and `DELETE` for orders,
* `CREATE`, `UPDATE` and `DELETE` for products, categories and suppliers.

[source, graphql, indent=0]
----
type Customer
    @node

>     @authentication( <1>

Reviewer: These points in the code are referenced with numbers, but I'm not sure what the reference to these are. Is this just to mark the code changes form the previous version?

Author: i think i was going to do something with them, but then did not.
removed :)

    supplier: [Supplier!]! @relationship(type: "SUPPLIES", direction: IN)
}

type Category
    @node
    @authentication( <4>
      operations: [CREATE, UPDATE, DELETE],
      jwt: { roles: { includes: "admin" } }
    ) {
    categoryName: String!
    products: [Product!]! @relationship(type: "PART_OF", direction: IN)
}

type Supplier
    @node
    @authentication( <5>
      operations: [CREATE, UPDATE, DELETE],
      jwt: { roles: { includes: "admin" } }
    ) {
    supplierID: ID! @id
    companyName: String!
    products: [Product!]! @relationship(type: "SUPPLIES", direction: OUT)
}
----

==== JSON Web Token (JWT) authentication

JWT authentication is a popular method for token-based authentication.
It allows clients to obtain and use tokens to authenticate subsequent requests.

JWT are represented by encoded JSON data.
These data can have arbitrary fields - which ones they should contain depends on the application preferences.

For instance, if the server side is trying to parse the `roles` field that was introduced in xref:#_authentication[], then the JWT should contain that.
With `@jwtClaim`, you can specify a path to a customer ID in a nested location.

> For example:

Reviewer: `@jwtClaim` is explained, but the directive `@jwt` is not mentioned anywhere, maybe worth mentioning that this directive allows to define the types of the jwt data?

}

type Category @node @authentication(operations: [CREATE, UPDATE, DELETE], jwt: { roles: { includes: "admin" } }) {
    categoryName: String!
    products: [Product!]! @relationship(type: "PART_OF", direction: IN)
}

type Supplier @node @authentication(operations: [CREATE, UPDATE, DELETE], jwt: { roles: { includes: "admin" } }) {
    supplierID: ID! @id
    companyName: String!
    products: [Product!]! @relationship(type: "SUPPLIES", direction: OUT)
}

type ordersProperties @relationshipProperties {
    unitPrice: Float!
    quantity: Int!
}
----

== Best practice checklist

Besides authentication and authorization considerations, there are a couple of worthwhile best practices to increase your API's security.

=== Avoid introspection and data field suggestions

While the xref:getting-started/graphql-aura.adoc[Getting started page for GraphQL and Aura Console] advocates to both **Enable introspection** as well as **Enable field suggestions**, this is not recommended when considering security.

Both potentially expose information that can be used to gain insight on specifics of your GraphQL schema and execute targeted malicious operations.

> Be sure to deactivate both in a customer-facing real-life scenario.

Reviewer: While this is definitely recommended in production systems. Introspection and field suggestions are not necessarily insecure, and some applications, such as public APIs may want these features, may be worth rewording this to make sure these are recommendations, but not critical steps?

Author: rephrased a bit

// Assume you have your schema definition in a string or file.
const typeDefs = readFileSync('./your-schema.graphql', 'utf-8');
const resolvers = { /* Your resolvers here. */ }; 
// Instantiate GraphQL Armor and configure the maxDepth plugin.
const armor = new ApolloArmor({
  maxDepth: {
    enabled: true,
    n: 5, // Sets the maximum allowed query depth to 5.
  },
});

// Get the security plugins provided by Armor.
const plugins = armor.protect();

const server = new ApolloServer({
  typeDefs,
  resolvers,
  plugins: [...plugins], // Add the armor plugins to Apollo Server.
});

const { url } = await startStandaloneServer(server, {
  listen: { port: 4000 },
});

console.log(`🚀 Server ready at ${url}`);
----

=== Paginate list fields

Returning large query result lists can negatively affect server performance.
For example, a query like the following would return a siginificant number of nodes:

[source, graphql, indent=0]
----
query {
  order(first: 1000) {
    orderID
    products(last: 100) {
      productName
      productCategory
    }
  }
}
----

You can prevent denial of service attacks based on queries such as this by paginating query results.

> A server-side pagination solution based on type definitions could look like this:

Reviewer: mmm, it feels like this example is missing the [`@limit`](https://neo4j.com/docs/graphql/current/directives/custom-logic/#_limit) directive that enforces limit

Or this is covering a different  point?

Author: where would you place it?

Reviewer: TBH, I'm not sure what this example is covering,  but if it is indeed server-side enforced pagination (https://neo4j.com/docs/graphql/current/directives/custom-logic/#_limit)

I'd say that we can add a `@limit` to any of the nodes of the initial typeDefinitions?

Author: added the `@limit` on the `Order` type definition

  pageInfo: PageInfo!
}

// The root query that is targeted
type Query {
  orders(first: Int, after: String, last: Int, before: String): OrderConnection
}
----

=== Rate-limit your API

Rate-limiting an API means setting an upper bound to how many requests a client can send in a certain amount of time or how costly those requests may be.
There is more than one approach.
Several are outlined in the following sections.

==== Query cost points

The link:https://shopify.dev/docs/api/usage/limits#the-leaky-bucket-algorithm[leaky bucket algorithm] represents an algorithmic solution to slow down the processing of multiple requests at once.

==== Query cost analysis

link:https://escape.tech/graphql-armor/docs/plugins/cost-limit/[GraphQL Armor] offers a way to limit the cost of GraphQL queries.

=== Use timeouts

To prevent the API from not responding or falling victim to denial of service attacks, it is feasible to make use of timeouts.
This way, subsequent queries aren't blocked by a long-running previous query.

You can set a timeout via the GraphQL Library driver, see xref:driver-configuration.adoc#_transaction_configuration_in_context[Transaction configuration in context].

> == Further reading

Reviewer: I would recommend reading the GraphQL security section: https://www.graphql.org/learn/security/

This covers up to date best practices for GraphQL in general

Author: added

[[securing-an-api]]
:description: This page is a tutorial on how to secure your API created with the Neo4j GraphQL Library.
= Securing your GraphQL API

This page is a tutorial on how to secure your API created with the Neo4j GraphQL Library.

== Prerequisites

. Set up a new AuraDB instance.
Refer to link:https://neo4j.com/docs/aura/getting-started/create-instance/[Creating a Neo4j Aura instance].
. Populate the instance with the Northwind data set.

[NOTE]
====
If you have completed the GraphQL and Aura Console getting started guide and would like to get rid of the example nodes you have created there, run the following in **Query** before populating your data base with the Northwind set:
[source,cypher]
----
MATCH (n) DETACH DELETE n;
----
====

> This tutorial builds on top of the xref:northwind-api.adoc[Northwind API tutorial].

Author: ```suggestion
This tutorial builds on top of the xref:graphql-modeling.adoc[GraphQL modeling tutorial].
```

    orders: [Product!]! @relationship(type: "ORDERS", direction: IN, properties: "ordersProperties")
    supplier: [Supplier!]! @relationship(type: "SUPPLIES", direction: IN)
}

type Category @node @authentication(operations: [CREATE, UPDATE, DELETE], jwt: { roles: { includes: "admin" } }) {
    categoryName: String!
    products: [Product!]! @relationship(type: "PART_OF", direction: IN)
}

type Supplier @node @authentication(operations: [CREATE, UPDATE, DELETE], jwt: { roles: { includes: "admin" } }) {
    supplierID: ID! @id
    companyName: String!
    products: [Product!]! @relationship(type: "SUPPLIES", direction: OUT)
}

type ordersProperties @relationshipProperties {
    unitPrice: Float!
    quantity: Int!
}
----

== Best practice checklist

Besides authentication and authorization considerations, there are a couple of worthwhile best practices to increase your API's security.

=== Avoid introspection and data field suggestions

While the xref:getting-started/graphql-aura.adoc[Getting started page for GraphQL and Aura Console] advocates to both **Enable introspection** as well as **Enable field suggestions**, this is not recommended when considering security.

> Both potentially expose information that can be used to gain insight on specifics of your GraphQL schema and execute targeted malicious opLorem ipsumerations.

Reviewer 2: ```suggestion
Both potentially expose information that can be used to gain insight on specifics of your GraphQL schema and execute targeted malicious operations.
```
possibly a typo?

Be sure to deactivate both in a customer-facing real-life scenario.

[[best-practice-internal-errors]]
=== Catch internal errors

For the same reason it is advisable to avoid introspection and data field suggestions, it can make your API more secure to catch internal errors and redact which information you want to pass on to the end user.

> For example, the following error reveals information XY:

Reviewer 2: Was this also a placeholder `XY`? feels like the JSON example here could be more illustrative?

[source, json, indent=0]
----
"data": {
  "field": "value"
}
----

You can use Apollo Server's link:https://www.apollographql.com/docs/apollo-server/data/errors[Error Handling] to catch such internal errors and then decide how to pass this on to your users:

[source, typescript, indent=0]
----
import { ApolloServerErrorCode } from '@apollo/server/errors';

if (error.extensions?.code === ApolloServerErrorCode.GRAPHQL_PARSE_FAILED) {
  // respond to the syntax error

} else if (error.extensions?.code === "MY_CUSTOM_CODE") {
  // do something else

}

> ----

Author: ```suggestion
```

=== Limit query depth

Limiting query depth disallows potentially harmful queries such as the following recursive query:

[source, graphql, indent=0]
----
query {
  order(id: 42) {
    products {
      order {
        products {
          order {
            products {
              order {
                # and so on...
              }
            }
          }
        }
      }
    }
  }
}
----

This can be achieved with a package such as link:https://www.npmjs.com/package/graphql-depth-limit[graphql-depth-limit]:

[source, typescript, indent=0]
----
import depthLimit from 'graphql-depth-limit'
import express from 'express'

> import graphqlHTTP from 'express-graphql'

Reviewer 2: should we maybe use a more up to date example here? we could replace the deprecated `express-graphql` with `graphql-http`, apparently. (cc: @mjfwebb)

Reviewer 3: We'll replace this example with an example using GraphQL Armor

import schema from './schema'
 
const app = express()
 
app.use('/graphql', graphqlHTTP((req, res) => ({
  schema,
  validationRules: [ depthLimit(10) ]
})))
----

=== Paginate list fields

Returning large query result lists can negatively affect server performance.
For example, a query like the following would return a siginificant number of nodes:

[source, graphql, indent=0]
----
query {
  order(first: 1000) {
    orderID
    products(last: 100) {
      productName
      productCategory
    }
  }
}
----

To avoid this, you can cap the input number directly in your resolvers, for example like this:

[source, graphql, indent=0]
----

> // example

Reviewer 2: missing example here I think?

----

Alternatively, use a library such as link:https://github.com/joonhocho/graphql-input-number[graphql-input-number]:

[source, typescript, indent=0]
----
import {
  GraphQLInputInt,
  GraphQLInputFloat,
} from 'graphql-input-number';

const argType = GraphQLInputInt({
  name: 'OneToNineInt',
  min: 1,
  max: 9,
});

new GraphQLObjectType({
  name: 'Query',
  fields: {
    input: {
      type: GraphQLInt,
      args: {
        number: {
          type: argType,
        },
      },
      resolve: (_, {number}) => {

        // 'number' IS AN INT BETWEEN 1 to 9.

      };
    },
  },
});

> ----

Author: replace with type definition solution (server side)
mention: denial of service prevention

=== Rate-limit your API

Rate-limiting an API means setting an upper bound to how many requests a client can send in a certain amount of time or how costly those requests may be.
There is more than one approach.
Several are outlined in the following sections.

==== Rate limit scores

Refer to GitHub's link:https://docs.github.com/en/rest/using-the-rest-api/rate-limits-for-the-rest-api?apiVersion=2022-11-28#primary-rate-limit-for-authenticated-users[Rate limits for the REST API].

==== Query cost points

The link:https://shopify.dev/docs/api/usage/limits#the-leaky-bucket-algorithm[leaky bucket algorithm].

==== Query cost analysis

> link:https://github.com/pa-bru/graphql-cost-analysis[graphql-cost-analysis]

Author: link to graphql armor

> link:https://github.com/pa-bru/graphql-cost-analysis[raphql-cost-analysis]

Reviewer 2: ```suggestion
link:https://github.com/pa-bru/graphql-cost-analysis[graphql-cost-analysis]
```

=== Use timeouts

To prevent the API from not responding or falling victim to denial of service attacks, it is feasible to make use of timeouts.
This way, subsequent queries will not be blocked by a long-running previous query.

There are many ways and places to use timeouts.
Here are a few examples.

> // examples

Author: refer to transaction config in context

> // examples

Reviewer 2: missing example?

=== Validate user input

User input may potentially be malicious, for example, it could contain code snippets which get executed when running queries against the database.

Follow the input validation methods summarized in the link:https://cheatsheetseries.owasp.org/cheatsheets/GraphQL_Cheat_Sheet.html#input-validation[OWASP Cheat Sheet Series].

// Examples?

> 

Author: ```suggestion
```

> // Examples?

Reviewer 2: also here, do we want an actual example?
